oauth2_http/java/com/google/auth/oauth2/JwtCredentials.java

```diff
@@ -0,0 +1,335 @@
+ *     .setSubject("some-subject@example.com")
+ *     .build();
+ * Credentials = JwtCredentials.newBuilder()
+ *     .setPrivateKey(privateKey)
+ *     .setPrivateKeyId("private-key-id")
+ *     .setClaims(claims)
+ *     .build();
+ * </code></pre>
+ */
+public class JwtCredentials extends Credentials implements JwtProvider {
+  private static final String JWT_ACCESS_PREFIX = OAuth2Utils.BEARER_PREFIX;
+  private static final String JWT_INCOMPLETE_ERROR_MESSAGE = "JWT claims must contain audience, "
+      + "issuer, and subject.";
+  private static final long CLOCK_SKEW = TimeUnit.MINUTES.toSeconds(5);
+
+  // byte[] is serializable, so the lock variable can be final
+  private final Object lock = new byte[0];
+  private final PrivateKey privateKey;
+  private final String privateKeyId;
+  private final Claims claims;
+  private final Long lifeSpanSeconds;
+  @VisibleForTesting
+  transient Clock clock;
+
+  private transient String jwt;
+  // The date (represented as seconds since the epoch) that the generated JWT expires
+  private transient Long expiryInSeconds;
+
+  JwtCredentials(Builder builder) {
+    this.privateKey = Preconditions.checkNotNull(builder.getPrivateKey());
+    this.privateKeyId = Preconditions.checkNotNull(builder.getPrivateKeyId());
+    this.claims = Preconditions.checkNotNull(builder.getClaims());
+    Preconditions.checkState(claims.isComplete(), JWT_INCOMPLETE_ERROR_MESSAGE);
+    this.lifeSpanSeconds = Preconditions.checkNotNull(builder.getLifeSpanSeconds());
+    this.clock = Preconditions.checkNotNull(builder.getClock());
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /**
+   * Refresh the token by discarding the cached token and metadata and rebuilding a new one.
+   */
+  @Override
+  public void refresh() throws IOException {
+    synchronized (lock) {
+      JsonWebSignature.Header header = new JsonWebSignature.Header();
+      header.setAlgorithm("RS256");
+      header.setType("JWT");
+      header.setKeyId(privateKeyId);
+
+      JsonWebToken.Payload payload = new JsonWebToken.Payload();
+      long currentTime = clock.currentTimeMillis();
+      payload.setAudience(claims.getAudience());
+      payload.setIssuer(claims.getIssuer());
+      payload.setSubject(claims.getSubject());
+      payload.setIssuedAtTimeSeconds(currentTime / 1000);
+      expiryInSeconds = currentTime / 1000 + lifeSpanSeconds;
+      payload.setExpirationTimeSeconds(expiryInSeconds);
+
+      JsonFactory jsonFactory = OAuth2Utils.JSON_FACTORY;
+
+      try {
+        jwt = JsonWebSignature.signUsingRsaSha256(privateKey, jsonFactory, header, payload);
+      } catch (GeneralSecurityException e) {
+        throw new IOException("Error signing service account JWT access header with private key.",
+            e);
+      }
+    }
+  }
+
+  private boolean shouldRefresh() {
+    return expiryInSeconds == null ||
+        getClock().currentTimeMillis() / 1000 > expiryInSeconds - CLOCK_SKEW;
```

Comment:
I'm not sure I'm following this but this doesn't look right. Can you elaborate? Why the skew? Why does it skew one way and not the other?

Reply:
The skew is to preemptively refresh credentials before they actually expire. If a host system's clock is different than Google's clock, then we might think our credentials are still good when they are not.

As for the logic here, it says we should refresh if we haven't generated a token yet or if our current clock says it's after 5 minutes before the expected expiry (so 5 minutes early)